### PR TITLE
Threadpoolexecutor + tqdm

### PR DESCRIPTION
**Pull request by ahartikainen**
#### Submission Checklist

- [x] Run unit tests
- [x] Declare copyright holder and open-source license: see below

#### Summary

Use tqdm (optional) for progressbar

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):

Ari Hartikainen

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

**Comment by ahartikainen** [comment]
This works, but for jupyter users, we need good instructions.

Is tqdm optional or should we add it to requirements.txt

What about tests?

There are many locations with update / refresh etc, this is due to how we gather info. If subprocess ends, it stops gathering input.

**Comment by ahartikainen** [comment]
There is still some cleanup left, so we don't need those multiple refresh rounds

**Comment by ahartikainen** [comment]
Current view in Terminal

output2

Current view in Jupyter Notebook

output

**Comment by mitzimorris** [comment]
Wow!!!
this is so cool!

**Comment by ahartikainen** [comment]
This is basically ready now. I added notes on the doc how to set-up this correctly. It is not enabled by default, so this is an optional feature.

I added convenience group for our optional dependencies

    pip install cmdstanpy[all]

which will install ujson and tqdm (both have wheels and should work on all the major platforms).

I don't have tests for this, and not sure what kind of tests we would want to have.

This is ready for the review.

**Comment by mitzimorris** [comment]
I will review this ASAP.   earlier today I checked out your branch and figured out tests to make sure that it runs the specified amount of threads in parallel - see   https://github.com/stan-dev/cmdstanpy/tree/ari/feature/threadpoolexecutor2

can this be merged back into your branch?

**Comment by ahartikainen** [comment]
I rebased based on your branch and also fixed some parts. Now everything should be working as intended.

File: cmdstanpy/model.py [hunk 1]
```diff
@@ -538,19 +612,16 @@ def run_generated_quantities(
                 data=_data,
                 seed=seed,
                 output_basename=gq_csv_basename,
-                method_args=generate_quantities_args
+                method_args=generate_quantities_args,
```
Comment (mitzimorris) [inline cmdstanpy/model.py]: don't need final comma  (nit-picky)

File: requirements.txt [hunk 2]
```diff
@@ -1,4 +1,2 @@
 pandas
 numpy>=1.15
-backports.tempfile==1.0
```
Comment (mitzimorris) [inline requirements.txt]: thanks for cleaning this up

File: cmdstanpy/model.py (hunk 1 above)
Comment (ahartikainen) [reply cmdstanpy/model.py]: I think that comes from `black`

Comment (mitzimorris) [reply cmdstanpy/model.py]: agreed - so it's fine to leave as is, and that's why I went ahead and merged.
I'd like to do another release to PyPi in order get parallelism back.

Comment (ahartikainen) [reply cmdstanpy/model.py]: New release sounds good

**Comment by mitzimorris** [comment]
I wish I'd tested more on my machine before merging - I'm seeing a problem with "show_progress=True" - not sure what's going on, but it triggers the following error:

```
Chain 4 - warmup: 1000it [00:00, 3194.68it/s]
Chain 4 - sample: 0it [00:00, ?it/s]
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-3-1e5e72fd5f66> in <module>
----> 1 fit = logistic_model.sample(data=logistic_data, chains=4, cores=4, csv_basename=os.path.join('.','foo2'), show_progress=True)

~/github/stan-dev/cmdstanpy/cmdstanpy/model.py in sample(self, data, chains, cores, seed, chain_ids, inits, warmup_iters, sampling_iters, save_warmup, thin, max_treedepth, metric, step_size, adapt_engaged, adapt_delta, csv_basename, show_progress)
    566                             msg, i, stanfit._retcode(i)
    567                         )
--> 568                 raise RuntimeError(msg)
    569             stanfit._validate_csv_files()
    570         return stanfit

RuntimeError: Error during sampling, chain 0 returned error code -1, chain 1 returned error code -1, chain 2 returned error code -1, chain 3 returned error code -1
```

note that tqdm on my machine isn't working properly - I'm seeing:

```
Chain 4 - warmup: 1000it [00:00, 3194.68it/s]
Chain 4 - sample: 0it [00:00, ?it/s]
```

do we need calls to `proc.wait` after all?

**Comment by ahartikainen** [comment]
I don't think this has anything to do with `proc.wait`

Problem is, that the real error message is hidden. Can you comment out RuntimeError and just call `raise`

**Comment by mitzimorris** [comment]
here's what got rid of the error:
moving call to set return_code to follow call to subprocess directly - plus proc.wait() - 
this is working on my machine:

```
 proc = subprocess.Popen(
            cmd.split(), stdout=subprocess.PIPE, stderr=subprocess.PIPE
        )
        proc.wait()
        stanfit._set_retcode(idx, proc.returncode)
        if pbar:
```

**Comment by ahartikainen** [comment]
What was the error?

**Comment by ahartikainen** [comment]
Ok, I can get this reproduce too

**Comment by mitzimorris** [comment]
so you agree, above fix is both good and necessary?

**Comment by mitzimorris** [comment]
we need a unit test to tickle this bug

**Comment by ahartikainen** [comment]
I'm not sure yet. Does pbar work normally.

I think this has something to do with threadpool executor, I'm currently testing.

**Comment by mitzimorris** [comment]
pbar isn't working at all on my machine.
is this an environment problem?

**Comment by mitzimorris** [comment]
this is what I'm seeing:
```
Chain 1 - warmup: 0it [00:00, ?it/s]
Chain 1 - sample: 0it [00:00, ?it/s]

Chain 2 - warmup: 0it [00:00, ?it/s]


Chain 2 - sample: 0it [00:00, ?it/s]



Chain 3 - warmup: 0it [00:00, ?it/s]




Chain 3 - sample: 0it [00:00, ?it/s]





Chain 4 - warmup: 0it [00:00, ?it/s]






Chain 1 - warmup: 0it [00:00, ?it/s]
Chain 1 - sample: 0it [00:00, ?it/s]
Chain 3 - warmup: 0it [00:00, ?it/s]
Chain 3 - sample: 0it [00:00, ?it/s]
Chain 2 - warmup: 0it [00:00, ?it/s]
Chain 2 - sample: 0it [00:00, ?it/s]
Chain 4 - warmup: 0it [00:00, ?it/s]
Chain 4 - sample: 0it [00:00, ?it/s]

​
```

**Comment by mitzimorris** [comment]
happy to let you take over this problem - please do this in a new branch.

**Comment by ahartikainen** [comment]
Ok, real error is 

```
Traceback (most recent call last):
  File "c:\users\user\github\cmdstanpy\cmdstanpy\model.py", line 766, in _run_cmdstan
    stdout_pbar = self._read_progress(proc, pbar)
  File "c:\users\user\github\cmdstanpy\cmdstanpy\model.py", line 728, in _read_progress
    raw_count - num_warmup - count_sampling,
    unsupported operand type(s) for -: 'int' and 'NoneType'
`` 

Where `print(raw_count, num_warmup, count_sampling) == 1001 None 0` after the proc.poll has ended.

**Comment by ahartikainen** [comment]
Hmm, `pbar.total` are saying `None`

**Comment by ahartikainen** [comment]
Oh, duh. I assumed that `sampler_args.warmup_iter` is always defined

```
                          tqdm_pbar(
                                desc="Chain {} - warmup".format(i + 1),
                                position=i * 2,
                                total=sampler_args.warmup_iters,
                                dynamic_ncols=True,
                            )
```

**Comment by ahartikainen** [comment]
Can we somehow get the default values for them?

**Comment by ahartikainen** [comment]
Ok, we can handle this dynamically.

So if those are None, let's read the total count from output and ... let's see how we can calculate the correct counts with basic info we have.

**Pull request merged**